{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/package/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88\n",
      "  return f(*args, **kwds)\n",
      "/package/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88\n",
      "  return f(*args, **kwds)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.lines import Line2D\n",
    "import seaborn as sns\n",
    "import scipy.stats as st\n",
    "import RELACS as rel\n",
    "sns.set_style(\"white\")\n",
    "sns.set_context(\"paper\")\n",
    "%load_ext rpy2.ipython\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NPC48h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bams: /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_iNPC/filtered_bam/iNPC_DMSO_rep1_H3K4me3_merged.filtered.bam /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_iNPC/filtered_bam/iNPC_DMSO_rep2_H3K4me3_merged.filtered.bam /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_iNPC/filtered_bam/iNPC_EPZ_rep1_H3K4me3_merged.filtered.bam /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_iNPC/filtered_bam/iNPC_EPZ_rep2_H3K4me3_merged.filtered.bam\n",
      "mark: H3K4me3\n",
      "outdir: ../INTERMEDIATE_FILES/NPC48h_DIFF_H3K4me3\n",
      "[1] TRUE\n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "[1] \"Intercept\"             \"rep_rep2_vs_rep1\"      \"condition_EPZ_vs_DMSO\"\n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "2019-10-24 15:46:38 [INFO] - Started UROPA\n",
      "2019-10-24 15:46:38 [INFO] - Working directory: /data/manke/group/ferrari/PhD_project/reproducible_code_paper/code_paper/figure_3/NOTEBOOKS\n",
      "2019-10-24 15:46:38 [INFO] - Command-line call: /home/ferrari/anaconda3/envs/uropa/bin/uropa -b ../INTERMEDIATE_FILES/NPC48h_DIFF_H3K4me3/peaks_Annotation.bed -g /home/ferrari/ferrari/my_repository/annotations_gencode/mouse/M18/annotation_snakePipes/gencode.vM18.annotation.sorted.gtf --feature transcript --distance 1500 --internals 1 -o ../INTERMEDIATE_FILES/NPC48h_DIFF_H3K4me3/output_uropa -t 10 --show_attributes gene_id transcript_id gene_name gene_type transcript_type\n",
      "2019-10-24 15:46:38 [INFO] - Reading configuration from commandline/input config\n",
      "2019-10-24 15:46:38 [INFO] - Reading .bed-file to annotate\n",
      "2019-10-24 15:46:38 [INFO] - Preparing .gtf-file for fast access\n",
      "2019-10-24 15:46:45 [INFO] - Started annotation\n",
      "2019-10-24 15:46:45 [INFO] - Progress: 0%\n",
      "2019-10-24 15:46:46 [INFO] - Progress: 9%\n",
      "2019-10-24 15:46:46 [INFO] - Progress: 20%\n",
      "2019-10-24 15:46:47 [INFO] - Progress: 30%\n",
      "2019-10-24 15:46:47 [INFO] - Progress: 42%\n",
      "2019-10-24 15:46:48 [INFO] - Progress: 53%\n",
      "2019-10-24 15:46:48 [INFO] - Progress: 69%\n",
      "2019-10-24 15:46:49 [INFO] - Progress: 82%\n",
      "2019-10-24 15:46:49 [INFO] - Progress: 92%\n",
      "2019-10-24 15:46:50 [INFO] - Progress: 100%\n",
      "2019-10-24 15:46:50 [INFO] - Processing annotated peaks\n",
      "2019-10-24 15:46:50 [INFO] - Writing output files\n",
      "2019-10-24 15:46:51 [INFO] - Cleaning up temporary files\n",
      "2019-10-24 15:46:51 [INFO] - UROPA run finished in 0:00:13!\n",
      "Namespace(AFTER_GENE=0, AFTER_TES=500, AFTER_TSS=500, BEFORE_GENE=0, BEFORE_TES=500, BEFORE_TSS=1000, FEATURE=['TSS'], PROTEIN_CODING=False, from_what='transcript', gtf_file='/home/ferrari/ferrari/my_repository/annotations_gencode/mouse/M18/annotation_snakePipes/gencode.vM18.annotation.sorted.gtf', out_dir='../INTERMEDIATE_FILES/NPC48h_DIFF_H3K4me3', verbose=False)\n",
      "creating final table ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Number of bins found: 15894\n",
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colMeans,\n",
      "    colnames, colSums, dirname, do.call, duplicated, eval, evalq,\n",
      "    Filter, Find, get, grep, grepl, intersect, is.unsorted, lapply,\n",
      "    lengths, Map, mapply, match, mget, order, paste, pmax, pmax.int,\n",
      "    pmin, pmin.int, Position, rank, rbind, Reduce, rowMeans, rownames,\n",
      "    rowSums, sapply, setdiff, sort, table, tapply, union, unique,\n",
      "    unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply\n",
      "\n",
      "converting counts to integer mode\n",
      "estimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates\n",
      "fitting model and testing\n",
      "using 'normal' for LFC shrinkage, the Normal prior from Love et al (2014).\n",
      "additional priors are available via the 'type' argument, see ?lfcShrink for details\n",
      "using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n",
      "\tPeak file = ../INTERMEDIATE_FILES/NPC48h_DIFF_H3K4me3/peaks_Annotation.bed\n",
      "\tGenome = mm10\n",
      "\tOrganism = mouse\n",
      "\tCustom annotation GTF file: /home/ferrari/ferrari/my_repository/annotations_gencode/mouse/M18/annotation_snakePipes/gencode.vM18.annotation.sorted.gtf (using transcript_id)\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 15595\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\t\tDuplicated Peak IDs: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 15595\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tReading Positions...\n",
      "\t-----------------------\n",
      "\tFinding Closest TSS...\n",
      "\tProcessing custom annotation file...\n",
      "\tFeatures that will be considered:\n",
      "\t\texon\n",
      "\n",
      "\t807591 lines contained useful feature information\n",
      "\tFeatures that will be considered:\n",
      "\t\texon\n",
      "\n",
      "\t807591 lines contained useful feature information\n",
      "\tPrioritizing Annotations: ......................\n",
      "\tAnnotating:......................\n",
      "\t\tAnnotation\tNumber of peaks\tTotal size (bp)\tLog2 Enrichment\n",
      "\t\tTTS\t355282.0\t95821634\t2.528\n",
      "\t\tExon\t488003.0\t80631665\t3.235\n",
      "\t\tIntron\t462666.0\t1008568397\t-0.487\n",
      "\t\tIntergenic\t0.0\t1446908478\t-20.740\n",
      "\t\tPromoter\t445766.0\t92359300\t2.908\n",
      "\tAnnotating:.....................\n",
      "\t\tAnnotation\tNumber of peaks\tTotal size (bp)\tLog2 Enrichment\n",
      "\t\tTTS\t547.0\t95819573\t-0.004\n",
      "\t\tExon\t2231.0\t80631665\t2.273\n",
      "\t\tIntron\t4695.0\t1008568397\t-0.298\n",
      "\t\tIntergenic\t749.0\t1446908478\t-3.467\n",
      "\t\tPromoter\t7373.0\t92345040\t3.802\n",
      "\tNOTE: If this part takes more than 2 minutes, there is a good chance\n",
      "\t\tyour machine ran out of memory: consider hitting ctrl+C and rerunning\n",
      "\t\tthe command with \"-noann\"\n",
      "\tTo capture annotation stats in a file, use \"-annStats <filename>\" next time\n",
      "\tAnnotating:\n",
      "\t\tAnnotation\tNumber of peaks\tTotal size (bp)\tLog2 Enrichment\n",
      "\tCounting Tags in Peaks from each directory...\n",
      "\tOrganism: mouse\n",
      "\tLoading Gene Informaiton...\n",
      "\tOutputing Annotation File...\n",
      "\tDone annotating peaks file\n",
      "\n",
      "sys:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "### PERFORM DESEQ2 DIFFERENTIAL ANALYSIS FOR EACH MARK ###\n",
    "# H3K4me3 #\n",
    "\n",
    "pipe_DiffPeaks.sh ../INPUT_DATA/info_data_H3K4me3_NPC48h.txt ../../figure_1-2/INPUT_DATA/NPC48h_H3K4me3_BAMPE-BAM_consensusPeaks.bed.cleaned.bed ../INTERMEDIATE_FILES/NPC48h_DIFF_H3K4me3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bams: /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_iNPC/filtered_bam/iNPC_DMSO_rep1_H3K27ac_merged.filtered.bam /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_iNPC/filtered_bam/iNPC_DMSO_rep2_H3K27ac_merged.filtered.bam /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_iNPC/filtered_bam/iNPC_EPZ_rep1_H3K27ac_merged.filtered.bam /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_iNPC/filtered_bam/iNPC_EPZ_rep2_H3K27ac_merged.filtered.bam\n",
      "mark: H3K27ac\n",
      "outdir: ../INTERMEDIATE_FILES/NPC48h_DIFF_H3K27ac\n",
      "[1] TRUE\n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "[1] \"Intercept\"             \"rep_rep2_vs_rep1\"      \"condition_EPZ_vs_DMSO\"\n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "2019-10-24 15:55:07 [INFO] - Started UROPA\n",
      "2019-10-24 15:55:07 [INFO] - Working directory: /data/manke/group/ferrari/PhD_project/reproducible_code_paper/code_paper/figure_3/NOTEBOOKS\n",
      "2019-10-24 15:55:07 [INFO] - Command-line call: /home/ferrari/anaconda3/envs/uropa/bin/uropa -b ../INTERMEDIATE_FILES/NPC48h_DIFF_H3K27ac/peaks_Annotation.bed -g /home/ferrari/ferrari/my_repository/annotations_gencode/mouse/M18/annotation_snakePipes/gencode.vM18.annotation.sorted.gtf --feature transcript --distance 1500 --internals 1 -o ../INTERMEDIATE_FILES/NPC48h_DIFF_H3K27ac/output_uropa -t 10 --show_attributes gene_id transcript_id gene_name gene_type transcript_type\n",
      "2019-10-24 15:55:07 [INFO] - Reading configuration from commandline/input config\n",
      "2019-10-24 15:55:07 [INFO] - Reading .bed-file to annotate\n",
      "2019-10-24 15:55:07 [INFO] - Preparing .gtf-file for fast access\n",
      "2019-10-24 15:55:14 [INFO] - Started annotation\n",
      "2019-10-24 15:55:14 [INFO] - Progress: 0%\n",
      "2019-10-24 15:55:15 [INFO] - Progress: 1%\n",
      "2019-10-24 15:55:15 [INFO] - Progress: 11%\n",
      "2019-10-24 15:55:16 [INFO] - Progress: 22%\n",
      "2019-10-24 15:55:16 [INFO] - Progress: 32%\n",
      "2019-10-24 15:55:17 [INFO] - Progress: 41%\n",
      "2019-10-24 15:55:17 [INFO] - Progress: 52%\n",
      "2019-10-24 15:55:18 [INFO] - Progress: 62%\n",
      "2019-10-24 15:55:18 [INFO] - Progress: 72%\n",
      "2019-10-24 15:55:19 [INFO] - Progress: 82%\n",
      "2019-10-24 15:55:19 [INFO] - Progress: 92%\n",
      "2019-10-24 15:55:20 [INFO] - Progress: 100%\n",
      "2019-10-24 15:55:20 [INFO] - Processing annotated peaks\n",
      "2019-10-24 15:55:20 [INFO] - Writing output files\n",
      "2019-10-24 15:55:21 [INFO] - Cleaning up temporary files\n",
      "2019-10-24 15:55:21 [INFO] - UROPA run finished in 0:00:14!\n",
      "Namespace(AFTER_GENE=0, AFTER_TES=500, AFTER_TSS=500, BEFORE_GENE=0, BEFORE_TES=500, BEFORE_TSS=1000, FEATURE=['TSS'], PROTEIN_CODING=False, from_what='transcript', gtf_file='/home/ferrari/ferrari/my_repository/annotations_gencode/mouse/M18/annotation_snakePipes/gencode.vM18.annotation.sorted.gtf', out_dir='../INTERMEDIATE_FILES/NPC48h_DIFF_H3K27ac', verbose=False)\n",
      "creating final table ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Number of bins found: 25450\n",
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colMeans,\n",
      "    colnames, colSums, dirname, do.call, duplicated, eval, evalq,\n",
      "    Filter, Find, get, grep, grepl, intersect, is.unsorted, lapply,\n",
      "    lengths, Map, mapply, match, mget, order, paste, pmax, pmax.int,\n",
      "    pmin, pmin.int, Position, rank, rbind, Reduce, rowMeans, rownames,\n",
      "    rowSums, sapply, setdiff, sort, table, tapply, union, unique,\n",
      "    unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply\n",
      "\n",
      "converting counts to integer mode\n",
      "estimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates\n",
      "fitting model and testing\n",
      "using 'normal' for LFC shrinkage, the Normal prior from Love et al (2014).\n",
      "additional priors are available via the 'type' argument, see ?lfcShrink for details\n",
      "using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n",
      "\tPeak file = ../INTERMEDIATE_FILES/NPC48h_DIFF_H3K27ac/peaks_Annotation.bed\n",
      "\tGenome = mm10\n",
      "\tOrganism = mouse\n",
      "\tCustom annotation GTF file: /home/ferrari/ferrari/my_repository/annotations_gencode/mouse/M18/annotation_snakePipes/gencode.vM18.annotation.sorted.gtf (using transcript_id)\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 22767\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\t\tDuplicated Peak IDs: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 22767\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tReading Positions...\n",
      "\t-----------------------\n",
      "\tFinding Closest TSS...\n",
      "\tProcessing custom annotation file...\n",
      "\tFeatures that will be considered:\n",
      "\t\texon\n",
      "\n",
      "\t807591 lines contained useful feature information\n",
      "\tFeatures that will be considered:\n",
      "\t\texon\n",
      "\n",
      "\t807591 lines contained useful feature information\n",
      "\tPrioritizing Annotations: ......................\n",
      "\tAnnotating:......................\n",
      "\t\tAnnotation\tNumber of peaks\tTotal size (bp)\tLog2 Enrichment\n",
      "\t\tTTS\t355242.0\t95821634\t2.527\n",
      "\t\tExon\t488046.0\t80627608\t3.235\n",
      "\t\tIntron\t462668.0\t1008572940\t-0.487\n",
      "\t\tIntergenic\t0.0\t1446908478\t-20.740\n",
      "\t\tPromoter\t445761.0\t92359300\t2.908\n",
      "\tAnnotating:.....................\n",
      "\t\tAnnotation\tNumber of peaks\tTotal size (bp)\tLog2 Enrichment\n",
      "\t\tTTS\t614.0\t95819573\t-0.383\n",
      "\t\tExon\t1799.0\t80627608\t1.417\n",
      "\t\tIntron\t7849.0\t1008572940\t-0.103\n",
      "\t\tIntergenic\t4729.0\t1446908478\t-1.354\n",
      "\t\tPromoter\t7776.0\t92345040\t3.333\n",
      "\tNOTE: If this part takes more than 2 minutes, there is a good chance\n",
      "\t\tyour machine ran out of memory: consider hitting ctrl+C and rerunning\n",
      "\t\tthe command with \"-noann\"\n",
      "\tTo capture annotation stats in a file, use \"-annStats <filename>\" next time\n",
      "\tAnnotating:\n",
      "\t\tAnnotation\tNumber of peaks\tTotal size (bp)\tLog2 Enrichment\n",
      "\tCounting Tags in Peaks from each directory...\n",
      "\tOrganism: mouse\n",
      "\tLoading Gene Informaiton...\n",
      "\tOutputing Annotation File...\n",
      "\tDone annotating peaks file\n",
      "\n",
      "sys:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "### PERFORM DESEQ2 DIFFERENTIAL ANALYSIS FOR EACH MARK ###\n",
    "# H3K27ac #\n",
    "\n",
    "pipe_DiffPeaks.sh ../INPUT_DATA/info_data_H3K27ac_NPC48h.txt ../../figure_1-2/INPUT_DATA/NPC48h_H3K27ac_BAMPE-BAM_consensusPeaks.bed.cleaned.bed ../INTERMEDIATE_FILES/NPC48h_DIFF_H3K27ac"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# mESC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bams: /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_mESC/filtered_bam/mESC_DMSO_rep1_H3K4me3_merged.filtered.bam /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_mESC/filtered_bam/mESC_DMSO_rep2_H3K4me3_merged.filtered.bam /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_mESC/filtered_bam/mESC_EPZ_rep1_H3K4me3_merged.filtered.bam /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_mESC/filtered_bam/mESC_EPZ_rep2_H3K4me3_merged.filtered.bam\n",
      "mark: H3K4me3\n",
      "outdir: ../INTERMEDIATE_FILES/mESC_DIFF_H3K4me3\n",
      "[1] TRUE\n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "[1] \"Intercept\"             \"rep_rep2_vs_rep1\"      \"condition_EPZ_vs_DMSO\"\n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "2019-10-24 16:02:44 [INFO] - Started UROPA\n",
      "2019-10-24 16:02:44 [INFO] - Working directory: /data/manke/group/ferrari/PhD_project/reproducible_code_paper/code_paper/figure_3/NOTEBOOKS\n",
      "2019-10-24 16:02:44 [INFO] - Command-line call: /home/ferrari/anaconda3/envs/uropa/bin/uropa -b ../INTERMEDIATE_FILES/mESC_DIFF_H3K4me3/peaks_Annotation.bed -g /home/ferrari/ferrari/my_repository/annotations_gencode/mouse/M18/annotation_snakePipes/gencode.vM18.annotation.sorted.gtf --feature transcript --distance 1500 --internals 1 -o ../INTERMEDIATE_FILES/mESC_DIFF_H3K4me3/output_uropa -t 10 --show_attributes gene_id transcript_id gene_name gene_type transcript_type\n",
      "2019-10-24 16:02:44 [INFO] - Reading configuration from commandline/input config\n",
      "2019-10-24 16:02:44 [INFO] - Reading .bed-file to annotate\n",
      "2019-10-24 16:02:44 [INFO] - Preparing .gtf-file for fast access\n",
      "2019-10-24 16:02:51 [INFO] - Started annotation\n",
      "2019-10-24 16:02:51 [INFO] - Progress: 0%\n",
      "2019-10-24 16:02:52 [INFO] - Progress: 10%\n",
      "2019-10-24 16:02:52 [INFO] - Progress: 19%\n",
      "2019-10-24 16:02:53 [INFO] - Progress: 24%\n",
      "2019-10-24 16:02:53 [INFO] - Progress: 30%\n",
      "2019-10-24 16:02:54 [INFO] - Progress: 40%\n",
      "2019-10-24 16:02:54 [INFO] - Progress: 50%\n",
      "2019-10-24 16:02:55 [INFO] - Progress: 57%\n",
      "2019-10-24 16:02:55 [INFO] - Progress: 63%\n",
      "2019-10-24 16:02:56 [INFO] - Progress: 71%\n",
      "2019-10-24 16:02:56 [INFO] - Progress: 80%\n",
      "2019-10-24 16:02:57 [INFO] - Progress: 86%\n",
      "2019-10-24 16:02:57 [INFO] - Progress: 95%\n",
      "2019-10-24 16:02:58 [INFO] - Progress: 100%\n",
      "2019-10-24 16:02:58 [INFO] - Processing annotated peaks\n",
      "2019-10-24 16:02:58 [INFO] - Writing output files\n",
      "2019-10-24 16:03:00 [INFO] - Cleaning up temporary files\n",
      "2019-10-24 16:03:00 [INFO] - UROPA run finished in 0:00:16!\n",
      "Namespace(AFTER_GENE=0, AFTER_TES=500, AFTER_TSS=500, BEFORE_GENE=0, BEFORE_TES=500, BEFORE_TSS=1000, FEATURE=['TSS'], PROTEIN_CODING=False, from_what='transcript', gtf_file='/home/ferrari/ferrari/my_repository/annotations_gencode/mouse/M18/annotation_snakePipes/gencode.vM18.annotation.sorted.gtf', out_dir='../INTERMEDIATE_FILES/mESC_DIFF_H3K4me3', verbose=False)\n",
      "creating final table ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Number of bins found: 28311\n",
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colMeans,\n",
      "    colnames, colSums, dirname, do.call, duplicated, eval, evalq,\n",
      "    Filter, Find, get, grep, grepl, intersect, is.unsorted, lapply,\n",
      "    lengths, Map, mapply, match, mget, order, paste, pmax, pmax.int,\n",
      "    pmin, pmin.int, Position, rank, rbind, Reduce, rowMeans, rownames,\n",
      "    rowSums, sapply, setdiff, sort, table, tapply, union, unique,\n",
      "    unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply\n",
      "\n",
      "converting counts to integer mode\n",
      "estimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates\n",
      "fitting model and testing\n",
      "using 'normal' for LFC shrinkage, the Normal prior from Love et al (2014).\n",
      "additional priors are available via the 'type' argument, see ?lfcShrink for details\n",
      "using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n",
      "\tPeak file = ../INTERMEDIATE_FILES/mESC_DIFF_H3K4me3/peaks_Annotation.bed\n",
      "\tGenome = mm10\n",
      "\tOrganism = mouse\n",
      "\tCustom annotation GTF file: /home/ferrari/ferrari/my_repository/annotations_gencode/mouse/M18/annotation_snakePipes/gencode.vM18.annotation.sorted.gtf (using transcript_id)\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 28197\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\t\tDuplicated Peak IDs: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 28197\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tReading Positions...\n",
      "\t-----------------------\n",
      "\tFinding Closest TSS...\n",
      "\tProcessing custom annotation file...\n",
      "\tFeatures that will be considered:\n",
      "\t\texon\n",
      "\n",
      "\t807591 lines contained useful feature information\n",
      "\tFeatures that will be considered:\n",
      "\t\texon\n",
      "\n",
      "\t807591 lines contained useful feature information\n",
      "\tPrioritizing Annotations: ......................\n",
      "\tAnnotating:......................\n",
      "\t\tAnnotation\tNumber of peaks\tTotal size (bp)\tLog2 Enrichment\n",
      "\t\tTTS\t355234.0\t95821634\t2.527\n",
      "\t\tExon\t488080.0\t80649274\t3.234\n",
      "\t\tIntron\t462615.0\t1008550517\t-0.487\n",
      "\t\tIntergenic\t0.0\t1446908478\t-20.740\n",
      "\t\tPromoter\t445788.0\t92359377\t2.908\n",
      "\tAnnotating:.....................\n",
      "\t\tAnnotation\tNumber of peaks\tTotal size (bp)\tLog2 Enrichment\n",
      "\t\tTTS\t1453.0\t95819573\t0.551\n",
      "\t\tExon\t3933.0\t80649274\t2.236\n",
      "\t\tIntron\t8064.0\t1008550517\t-0.372\n",
      "\t\tIntergenic\t5191.0\t1446908478\t-1.529\n",
      "\t\tPromoter\t9556.0\t92345040\t3.322\n",
      "\tNOTE: If this part takes more than 2 minutes, there is a good chance\n",
      "\t\tyour machine ran out of memory: consider hitting ctrl+C and rerunning\n",
      "\t\tthe command with \"-noann\"\n",
      "\tTo capture annotation stats in a file, use \"-annStats <filename>\" next time\n",
      "\tAnnotating:\n",
      "\t\tAnnotation\tNumber of peaks\tTotal size (bp)\tLog2 Enrichment\n",
      "\tCounting Tags in Peaks from each directory...\n",
      "\tOrganism: mouse\n",
      "\tLoading Gene Informaiton...\n",
      "\tOutputing Annotation File...\n",
      "\tDone annotating peaks file\n",
      "\n",
      "sys:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "### PERFORM DESEQ2 DIFFERENTIAL ANALYSIS FOR EACH MARK ###\n",
    "# H3K4me3 #\n",
    "\n",
    "pipe_DiffPeaks.sh ../INPUT_DATA/info_data_H3K4me3_mESC.txt ../../figure_1-2/INPUT_DATA/mESC_H3K4me3_BAMPE-BAM_consensusPeaks.bed.cleaned.bed ../INTERMEDIATE_FILES/mESC_DIFF_H3K4me3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bams: /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_mESC/filtered_bam/mESC_DMSO_rep1_H3K27ac_merged.filtered.bam /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_mESC/filtered_bam/mESC_DMSO_rep2_H3K27ac_merged.filtered.bam /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_mESC/filtered_bam/mESC_EPZ_rep1_H3K27ac_merged.filtered.bam /data/manke/group/ferrari/PhD_project/reference_datasets/Ferrari_mESC-iNPC_DMSOvsEPZ_fullEpigenomes_MERGED/output_DNA-mapping_snakepipe_mESC/filtered_bam/mESC_EPZ_rep2_H3K27ac_merged.filtered.bam\n",
      "mark: H3K27ac\n",
      "outdir: ../INTERMEDIATE_FILES/mESC_DIFF_H3K27ac\n",
      "[1] TRUE\n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "[1] \"Intercept\"             \"rep_rep2_vs_rep1\"      \"condition_EPZ_vs_DMSO\"\n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "2019-10-24 16:10:45 [INFO] - Started UROPA\n",
      "2019-10-24 16:10:45 [INFO] - Working directory: /data/manke/group/ferrari/PhD_project/reproducible_code_paper/code_paper/figure_3/NOTEBOOKS\n",
      "2019-10-24 16:10:45 [INFO] - Command-line call: /home/ferrari/anaconda3/envs/uropa/bin/uropa -b ../INTERMEDIATE_FILES/mESC_DIFF_H3K27ac/peaks_Annotation.bed -g /home/ferrari/ferrari/my_repository/annotations_gencode/mouse/M18/annotation_snakePipes/gencode.vM18.annotation.sorted.gtf --feature transcript --distance 1500 --internals 1 -o ../INTERMEDIATE_FILES/mESC_DIFF_H3K27ac/output_uropa -t 10 --show_attributes gene_id transcript_id gene_name gene_type transcript_type\n",
      "2019-10-24 16:10:45 [INFO] - Reading configuration from commandline/input config\n",
      "2019-10-24 16:10:45 [INFO] - Reading .bed-file to annotate\n",
      "2019-10-24 16:10:45 [INFO] - Preparing .gtf-file for fast access\n",
      "2019-10-24 16:10:52 [INFO] - Started annotation\n",
      "2019-10-24 16:10:53 [INFO] - Progress: 0%\n",
      "2019-10-24 16:10:54 [INFO] - Progress: 10%\n",
      "2019-10-24 16:10:54 [INFO] - Progress: 13%\n",
      "2019-10-24 16:10:54 [INFO] - Progress: 14%\n",
      "2019-10-24 16:10:55 [INFO] - Progress: 20%\n",
      "2019-10-24 16:10:55 [INFO] - Progress: 30%\n",
      "2019-10-24 16:10:56 [INFO] - Progress: 38%\n",
      "2019-10-24 16:10:56 [INFO] - Progress: 43%\n",
      "2019-10-24 16:10:57 [INFO] - Progress: 51%\n",
      "2019-10-24 16:10:57 [INFO] - Progress: 60%\n",
      "2019-10-24 16:10:58 [INFO] - Progress: 64%\n",
      "2019-10-24 16:10:58 [INFO] - Progress: 73%\n",
      "2019-10-24 16:10:59 [INFO] - Progress: 81%\n",
      "2019-10-24 16:10:59 [INFO] - Progress: 90%\n",
      "2019-10-24 16:11:00 [INFO] - Progress: 94%\n",
      "2019-10-24 16:11:00 [INFO] - Progress: 100%\n",
      "2019-10-24 16:11:01 [INFO] - Processing annotated peaks\n",
      "2019-10-24 16:11:01 [INFO] - Writing output files\n",
      "2019-10-24 16:11:02 [INFO] - Cleaning up temporary files\n",
      "2019-10-24 16:11:02 [INFO] - UROPA run finished in 0:00:17!\n",
      "Namespace(AFTER_GENE=0, AFTER_TES=500, AFTER_TSS=500, BEFORE_GENE=0, BEFORE_TES=500, BEFORE_TSS=1000, FEATURE=['TSS'], PROTEIN_CODING=False, from_what='transcript', gtf_file='/home/ferrari/ferrari/my_repository/annotations_gencode/mouse/M18/annotation_snakePipes/gencode.vM18.annotation.sorted.gtf', out_dir='../INTERMEDIATE_FILES/mESC_DIFF_H3K27ac', verbose=False)\n",
      "creating final table ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Number of bins found: 30458\n",
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colMeans,\n",
      "    colnames, colSums, dirname, do.call, duplicated, eval, evalq,\n",
      "    Filter, Find, get, grep, grepl, intersect, is.unsorted, lapply,\n",
      "    lengths, Map, mapply, match, mget, order, paste, pmax, pmax.int,\n",
      "    pmin, pmin.int, Position, rank, rbind, Reduce, rowMeans, rownames,\n",
      "    rowSums, sapply, setdiff, sort, table, tapply, union, unique,\n",
      "    unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply\n",
      "\n",
      "converting counts to integer mode\n",
      "estimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates\n",
      "fitting model and testing\n",
      "using 'normal' for LFC shrinkage, the Normal prior from Love et al (2014).\n",
      "additional priors are available via the 'type' argument, see ?lfcShrink for details\n",
      "using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n",
      "\tPeak file = ../INTERMEDIATE_FILES/mESC_DIFF_H3K27ac/peaks_Annotation.bed\n",
      "\tGenome = mm10\n",
      "\tOrganism = mouse\n",
      "\tCustom annotation GTF file: /home/ferrari/ferrari/my_repository/annotations_gencode/mouse/M18/annotation_snakePipes/gencode.vM18.annotation.sorted.gtf (using transcript_id)\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 30448\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\t\tDuplicated Peak IDs: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 30448\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tReading Positions...\n",
      "\t-----------------------\n",
      "\tFinding Closest TSS...\n",
      "\tProcessing custom annotation file...\n",
      "\tFeatures that will be considered:\n",
      "\t\texon\n",
      "\n",
      "\t807591 lines contained useful feature information\n",
      "\tFeatures that will be considered:\n",
      "\t\texon\n",
      "\n",
      "\t807591 lines contained useful feature information\n",
      "\tPrioritizing Annotations: ......................\n",
      "\tAnnotating:......................\n",
      "\t\tAnnotation\tNumber of peaks\tTotal size (bp)\tLog2 Enrichment\n",
      "\t\tTTS\t355241.0\t95821634\t2.527\n",
      "\t\tExon\t488055.0\t80652586\t3.234\n",
      "\t\tIntron\t462653.0\t1008558175\t-0.487\n",
      "\t\tIntergenic\t0.0\t1446908478\t-20.740\n",
      "\t\tPromoter\t445768.0\t92359231\t2.908\n",
      "\tAnnotating:.....................\n",
      "\t\tAnnotation\tNumber of peaks\tTotal size (bp)\tLog2 Enrichment\n",
      "\t\tTTS\t1366.0\t95819573\t0.351\n",
      "\t\tExon\t2149.0\t80652586\t1.253\n",
      "\t\tIntron\t9831.0\t1008558175\t-0.197\n",
      "\t\tIntergenic\t7983.0\t1446908478\t-1.018\n",
      "\t\tPromoter\t9119.0\t92345040\t3.143\n",
      "\tNOTE: If this part takes more than 2 minutes, there is a good chance\n",
      "\t\tyour machine ran out of memory: consider hitting ctrl+C and rerunning\n",
      "\t\tthe command with \"-noann\"\n",
      "\tTo capture annotation stats in a file, use \"-annStats <filename>\" next time\n",
      "\tAnnotating:\n",
      "\t\tAnnotation\tNumber of peaks\tTotal size (bp)\tLog2 Enrichment\n",
      "\tCounting Tags in Peaks from each directory...\n",
      "\tOrganism: mouse\n",
      "\tLoading Gene Informaiton...\n",
      "\tOutputing Annotation File...\n",
      "\tDone annotating peaks file\n",
      "\n",
      "sys:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "### PERFORM DESEQ2 DIFFERENTIAL ANALYSIS FOR EACH MARK ###\n",
    "# H3K27ac #\n",
    "\n",
    "pipe_DiffPeaks.sh ../INPUT_DATA/info_data_H3K27ac_mESC.txt ../../figure_1-2/INPUT_DATA/mESC_H3K27ac_BAMPE-BAM_consensusPeaks.bed.cleaned.bed ../INTERMEDIATE_FILES/mESC_DIFF_H3K27ac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": " Python 3.6 ipython@maximus (MPI-IE)",
   "language": "",
   "name": "rik_ssh_maximus_python36ipythonmaximusmpiie"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
